capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1424 @@
+                    recv_len: usize) {
+        // Note: a good default recv_len is 10 bytes. Reading too many bytes
+        //  rarely matters. However, it occasionally matters a lot, so we do
+        //  provided a settable recv_len
+
+        if self.is_initialized() {
+            // device is already initialized
+            self.set_spi_fast_mode();
+        } else {
+            // device is still being initialized
+            self.set_spi_slow_mode();
+        }
+
+        // check that write buffer is long enough
+        if write_buffer.len() < 8 + recv_len {
+            panic!("Write buffer too short to send SD card commands");
+        }
+
+        // send dummy bytes to start
+        write_buffer[0] = 0xFF;
+        write_buffer[1] = 0xFF;
+
+        // command
+        if (0x80 & cmd as u8) != 0x00 {
+            // application-specific command
+            write_buffer[2] = 0x40 | (0x7F & cmd as u8);
+        } else {
+            // normal command
+            write_buffer[2] = 0x40 | cmd as u8;
+        }
+
+        // argument, MSB first
+        write_buffer[3] = ((arg >> 24) & 0xFF) as u8;
+        write_buffer[4] = ((arg >> 16) & 0xFF) as u8;
+        write_buffer[5] = ((arg >> 8) & 0xFF) as u8;
+        write_buffer[6] = ((arg >> 0) & 0xFF) as u8;
+
+        // CRC is ignored except for CMD0 and maybe CMD8
+        if cmd == SDCmd::CMD8 {
+            write_buffer[7] = 0x87; // valid crc for CMD8(0x1AA)
+        } else {
+            write_buffer[7] = 0x95; // valid crc for CMD0
+        }
+
+        // append dummy bytes to transmission
+        for i in 0..recv_len {
+            write_buffer[8 + i] = 0xFF;
+        }
+
+        self.spi.read_write_bytes(write_buffer, Some(read_buffer), 8 + recv_len);
+    }
+
+    fn read_bytes(&self,
+                  mut write_buffer: &'static mut [u8],
+                  mut read_buffer: &'static mut [u8],
+                  recv_len: usize) {
+
+        self.set_spi_fast_mode();
+
+        // set write buffer to null transactions
+        // Note: this could be optimized in the future by allowing SPI to read
+        //  without a write buffer passed in
+        let count = cmp::min(write_buffer.len(), recv_len);
+        for i in 0..count {
+            write_buffer[i] = 0xFF;
+        }
+
+        self.spi.read_write_bytes(write_buffer, Some(read_buffer), recv_len);
+    }
+
+    fn write_bytes(&self,
+                   mut write_buffer: &'static mut [u8],
+                   mut read_buffer: &'static mut [u8],
+                   recv_len: usize) {
+
+        self.set_spi_fast_mode();
+
+        self.spi.read_write_bytes(write_buffer, Some(read_buffer), recv_len);
+    }
+
+
+    fn get_response(&self, response: SDResponse, read_buffer: &[u8]) -> (u8, u8, u32) {
+
+        let mut r1: u8 = 0xFF;
+        let mut r2: u8 = 0xFF;
+        let mut r3: u32 = 0xFFFFFFFF;
+
+        // scan through read buffer for response byte
+        for i in 0..read_buffer.len() {
```

Comment:
It's a little more awkward because of the `i+n`'s in the body, but I'm suspecting there's something analogous to python's `enumerate` that would at least provide some iteration safety as well as the current index?